### PR TITLE
magento/magento2#35934: Internal Server Error in special-price API Endpoint

### LINKED ISSUE #35934: Internal Server Error in /V1/products/special-price API Endpoint
### Preconditions and environment

- Magento version 2.4

### Steps to reproduce

1. Retrieve auth token from the admin endpoint /V1/integration/admin/
2. Send post request to below endpoints with the payload: 
Payload:
_{"prices": "[]"}_

or
_{"prices": ""}_

API's effected with above payload:

1. POST request to http://yourstore/rest/default/V1/products/special-price
2. /V1/products/base-prices
3. /V1/products/cost
4. /V1/products/special-price-delete
5. /V1/products/tier-prices-delete
6. /V1/products/tier-prices
7. /V1/inventory/is-product-salable-for-requested-qty/{sku}/{stockId}/{requestedQty}

### Expected result

Returns a 400 error as the request is malformed.

### Actual result

Returns a 500 status code from a null TypeError .

### Additional information

output of var/log/system.log:
```
	[2022-08-01T13:05:39.434987+00:00] main.CRITICAL: TypeError: Magento\Catalog\Model\Product\Price\SpecialPriceStorage::update(): Argument #1 ($prices) must be of type array, null given in /magento/vendor/magento/module-catalog/Model/Product/Price/SpecialPriceStorage.php:125
	Stack trace:
	#0 [internal function]: Magento\Catalog\Model\Product\Price\SpecialPriceStorage->update()
	#1 /magento/vendor/magento/module-webapi/Controller/Rest/SynchronousRequestProcessor.php(95): call_user_func_array()
	#2 /magento/vendor/magento/module-webapi/Controller/Rest.php(188): Magento\Webapi\Controller\Rest\SynchronousRequestProcessor->process()
	#3 /magento/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Webapi\Controller\Rest->dispatch()
	#4 /magento/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Webapi\Controller\Rest\Interceptor->___callParent()
	#5 /magento/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Webapi\Controller\Rest\Interceptor->Magento\Framework\Interception\{closure}()
	#6 /magento/generated/code/Magento/Webapi/Controller/Rest/Interceptor.php(23): Magento\Webapi\Controller\Rest\Interceptor->___callPlugins()
	#7 /magento/vendor/magento/framework/App/Http.php(116): Magento\Webapi\Controller\Rest\Interceptor->dispatch()
	#8 /magento/generated/code/Magento/Framework/App/Http/Interceptor.php(23): Magento\Framework\App\Http->launch()
	#9 /magento/vendor/magento/framework/App/Bootstrap.php(264): Magento\Framework\App\Http\Interceptor->launch()
	#10 /magento/pub/index.php(30): Magento\Framework\App\Bootstrap->run()
	#11 {main} [] []
```

### Release note

_No response_

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [X] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by serbynskyi**
### Fixed Issues (if relevant)

1. Fixes magento/magento2#35934

### Manual testing scenarios (*)
1. Retrieve auth token from the admin endpoint /V1/integration/admin/
2. Send post request to below endpoints with the payload:
Payload:
{"prices": "[]"}

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by serbynskyi** [comment]
@magento run all tests

**Comment by serbynskyi** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/Webapi/ServiceInputProcessor.php [hunk 1]
```diff
@@ -519,6 +520,12 @@ public function convertValue($data, $type)
                     foreach ($data as $key => $item) {
                         $result[$key] = $this->_createFromArray($itemType, $item);
                     }
+                } else {
+                    throw new SerializationException(
+                        new Phrase(
+                            'The value "' . $data . '" must be an array.'
+                        )
```
Comment (hostep) [inline lib/internal/Magento/Framework/Webapi/ServiceInputProcessor.php:528]: Injecting a dynamic value in a translatable string will make it very hard to translate it.
I would suggest doing something like this instead:

`new Phrase('The value "%fieldName" must be an array.', ['fieldName' => $data]);`

**Comment by serbynskyi** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/Webapi/ServiceInputProcessor.php (hunk 1 above)
Comment (serbynskyi) [reply lib/internal/Magento/Framework/Webapi/ServiceInputProcessor.php:528]: Yes, I agree. The pull request is updated.

**Comment by serbynskyi** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/Webapi/ServiceInputProcessor.php [hunk 2]
```diff
@@ -519,6 +520,13 @@ public function convertValue($data, $type)
                     foreach ($data as $key => $item) {
                         $result[$key] = $this->_createFromArray($itemType, $item);
                     }
+                } else {
```
Comment (engcom-Hotel) [inline lib/internal/Magento/Framework/Webapi/ServiceInputProcessor.php:523]: Please add some automated tests with a negative scenario where we are not passing the array.

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Hello @serbynskyi,

Thanks for the contribution!

Please fix the Integration and web API test failures.

Thanks

**Comment by engcom-Echo** [comment]
@serbynskyi 
Thank you for your contribution!
While going through the PR, I have found that original scope of the issue was to fix for special-prices but seems the fix is causing regression issue. 
Can you please have a look at the [integration](https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/37294/3212539b087f18a6eb01db7713fb28d6/Integration/console-error-logs.html) and [Web API](https://public-results-storage-prod.magento-testing-service.engineering/reports/magento/magento2/pull/37294/1567502496cdf95eecf0573bbbe24481/WebApi/console-error-logs.html) test failures those seems to be due the PR changes.

In Order to move PR further build should be green.

**Comment by engcom-Echo** [comment]
@serbynskyi we are moving this PR `On Hold` based on [comment](https://github.com/magento/magento2/pull/37294#issuecomment-1559293865). Let us know when you are ready to work on it again, and we will be happy to move it back further. Thank you for your contribution.

**Comment by engcom-Echo** [comment]
Since it needs changes as per above [comment](https://github.com/magento/magento2/pull/37294#issuecomment-1559293865) maving it to Draft state.

**Comment by engcom-Echo** [comment]
@serbynskyi Since it needs changes as per above [comment](https://github.com/magento/magento2/pull/37294#issuecomment-1559293865) maving it to Draft state.

**Comment by engcom-Echo** [comment]
As mentioned in above [comment](https://github.com/magento/magento2/pull/37294#issuecomment-1559293865) since there are regression issues caused by PR changes as we as there are some build failures closing this PR for now.

Please feel free to reopen for any update.

Thank you
